elu function implementation

samthakur587 opened the pull request:
Close #17456

samthakur587 commented:
#17558 this the issue for this pull

samthakur587 commented:
it's ready to review the pull

ivy-leaves commented:
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

samthakur587 commented:
hey @hirwa-nshuti there is so much merge conflict happening in pull request so can i make a new pull request with a new branch

hirwa-nshuti reviewed (changes requested):
Hey @samthakur587 

Would you resolve the merge conflicts for CI run

Thanks

samthakur587 commented:
Hey @hirwa-nshuti 

I have changed only the 6 file for my PR but to solve the merge conflict the 87 files had changed that are not related to my PR and not it not solving from my side can i make a new PR from a New branch to solve all the issue.

Thank you

samthakur587 commented:
finally all previous commit was reset @hirwa-nshuti  now you can go ahead and review my PR 

thanks

hirwa-nshuti commented on ivy/functional/ivy/activations.py:
```diff
@@ -609,3 +610,69 @@ def mish(
+    x
+        Input array.
+    alpha
+        sclaer multiple of exponent slop of negative input.
```
There is a typo here in scalar and slope would you look into the rest of docstring

samthakur587 replied:
i have corrected this typo in doc string you can check it now.

thanks

hirwa-nshuti reviewed (changes requested):
Hey would you add the issue number of subtask that you're working on

samthakur587 commented:
the issue number of subtask that i am  working on is Extend Ivy Functional API #3856

hirwa-nshuti commented:
> the issue number of subtask that i am working on is Extend Ivy Functional API #3856

Would you refer to contributing guide and walk through the way to assign yourself a task https://unify.ai/docs/ivy/overview/contributing/open_tasks.html#open-tasks

samthakur587 commented:
sorry for the mistake i have added the PR number in place of issue number in PR description. Now i edited the PR description with issue number. you can check it.
 
thank you

hirwa-nshuti reviewed (changes requested):
Hey,

The tests aren't passing in either backends, also you'll have to impelement everything in the experimental API
https://github.com/unifyai/ivy/tree/master/ivy/functional/ivy/experimental

samthakur587 commented:
hey @hirwa-nshuti previously i was implemented the elu activation function outside the ivy  experimental API now i removed all the method from there and implemented with in ivy experimental api you can check now.

thank you

samthakur587 commented:
![image](https://github.com/unifyai/ivy/assets/83540902/b61037d5-e246-4f8c-8e16-8626e5d13e24)

this test of the activation function silu and its not related to my PR. why it is coming i don't know can you please check why it is coming. what changes do i required to solve this.

thank you

samthakur587 commented:
hey @hirwa-nshuti completed all the changes . now its ready to merge

samthakur587 commented on ivy_tests/array_api_testing/test_array_api:
all the changes are completed. now you can review.

hirwa-nshuti commented on ivy/functional/backends/tensorflow/experimental/activations.py:
```diff
@@ -66,3 +66,12 @@ def silu(
+@with_supported_dtypes({"2.12.0 and below": ("float",)}, backend_version)
+def elu(x: Tensor, /, *, alpha: float = 1.0, out: Optional[Tensor] = None) -> Tensor:
+    alpha = tf.cast(alpha, x.dtype)
+    ret = tf.cast(tf.where(x > 0, x, alpha * tf.expm1(x)), x.dtype)
```
This line is breaking the tests as we call `tf.math.expm1` not `tf.expm1`

samthakur587 replied:
hey @hirwa-nshuti  i have changed the tf.expm1 to tf.math.expm1 and also implemented the elu function for paddle backend. in the paddle backend i called the elu function from paddle.nn.function class so i implemented the elu function in paddle.nn.functional as well with its test function. you can check and if there any mistake please let me know.

thank you

hirwa-nshuti reviewed (changes requested):
Hey some minor changes to get the PR merged and also you need to add paddle backend implementation of this function.

Thanks😊

hirwa-nshuti commented on ivy/functional/frontends/paddle/nn/functional/activation.py:
```diff
@@ -193,6 +193,20 @@ def relu_(x, name=None):
     return x
 
 
+@with_supported_dtypes({"2.4.2 and below": ("float32", "float64")}, "paddle")
```
Here we need to call `ivy.elu` as we are implementing the same function

hirwa-nshuti reviewed (changes requested):
Hey good job so far,
Just run your tests locally and they're failing for the paddle backend, due to assertion error, would you look into this😊
And check the comments below

samthakur587 commented:
![Screenshot 2023-06-28 173223](https://github.com/unifyai/ivy/assets/83540902/d1f5f4b3-a035-4568-9699-5e19962387e3)

Hey @hirwa-nshuti i am trying to test the paddle frontend test in local i am getting the keyerror  'cpu' i don't where it is store and why this error is coming. can you please help me come out from this problem.

the command i run is : -  
pytest -vv ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_paddle_activation.py::test_paddle_elu

thank you

hirwa-nshuti commented on ivy/functional/frontends/paddle/nn/functional/activation.py:
```diff
@@ -193,6 +193,18 @@ def relu_(x, name=None):
     return x
 
 
+@with_supported_dtypes({"2.4.2 and below": ("float32", "float64")}, "paddle")
```
We now use the paddle latest version `2.5.0`

samthakur587 replied:
done , changed the paddle version to 2.5.0

hirwa-nshuti reviewed (changes requested):
With the comment I added will fix the tests for four backends and then you'll need to fix value test error in the paddle backend implementation of the function🙂

samthakur587 commented:
hey @hirwa-nshuti i have run all the tests locally all are passing but one warring has encounter from numpy backed expm1 function that is show here . 

![image](https://github.com/unifyai/ivy/assets/83540902/135191d9-b84c-4eae-8af0-79bf8671e9be)

do you has any suggestion to overcome from this. otherwise the PR was completed and ready to merge.

thank you

hirwa-nshuti reviewed (changes requested):
if you run `pytest -v ivy_tests/test_ivy/test_functional/te
st_experimental/test_nn/test_activations.py::test_elu` 
there is failure for paddle backend other changes looks great, just a fix is needed for the paddle backend

samthakur587 commented:
![image](https://github.com/unifyai/ivy/assets/83540902/ce0753bc-48ec-40ef-8c34-1b7ac9e55c76)

Finally all tests are passing locally. Now its Ready to merge . 

thank you

hirwa-nshuti commented on ivy/functional/frontends/paddle/nn/functional/activation.py:
```diff
@@ -193,6 +193,18 @@ def relu_(x, name=None):
+    alpha=1.0,
+    name=None,
+):
+    return ivy.elu(x, alpha)
```
```suggestion
    return ivy.elu(x, alpha=alpha)
```

hirwa-nshuti reviewed (approved):
Hey @samthakur587 

All looks good now, and I'm merging this PR😅
Thanks for contributing😊

samthakur587 commented:
Hey @hirwa-nshuti 

That's awesome ! I really appreciate all the help and guidance you've provided throughout this process. It wouldn't have been possible without your valuable contributions.

 Thanks a ton for your support! 😄
